vendor/k8s.io/kubernetes/pkg/kubelet/kubelet.go

```diff
@@ -2105,16 +2107,21 @@ func (kl *Kubelet) cleanupOrphanedPodDirs(
 		if allPods.Has(string(uid)) {
 			continue
 		}
+		// If volumes have not been unmounted/detached, do not delete directory.
+		// Doing so may result in corruption of data.
 		if podVolumesExist := kl.podVolumesExist(uid); podVolumesExist {
-			// If volumes have not been unmounted/detached, do not delete directory.
-			// Doing so may result in corruption of data.
 			glog.V(3).Infof("Orphaned pod %q found, but volumes are not cleaned up; err: %v", uid, err)
 			continue
 		}
+		// Check whether volume is still mounted on disk. If so, do not delete directory
+		if volumeNames, err := kl.getPodVolumeNameListFromDisk(uid); err != nil || len(volumeNames) != 0 {
```

Comment:
@eparis @jsafrane so I reviewed this PR in some detail and this is the bit - which helps with volumes not getting deleted if it is still mounted on disk. 

The other big change here is in reconciler.go:

1. Refactor `reconcile()` function in its own function. There isn't code a code change in whole of reconcilation itself but diff looks pretty big because of refactoring.
2. After reconcillation - reconciler also periodically calls `reconstruct` - which populates state of world from pod (and iterating through each pod volume plugin directory)  and mounts as necessary.  

Overall, I think we should be okay.